File: bind/ts/src/events/pointer.ts
import type { Vector2, ElementState, TouchPhase } from './common';

// MARK: Pointer Events

/** Mouse button types */
export type MouseButton =
  | 'left'
  | 'right'
  | 'middle'
  | 'back'
  | 'forward'
  | { other: number };

/** Pointer type for unified mouse/touch handling */
export type PointerType = 'mouse' | 'touch' | 'pen';

/** Mouse scroll delta type */
export type ScrollDelta =
  | { type: 'line'; value: Vector2 }
  | { type: 'pixel'; value: Vector2 };

/**
 * Event fired when the pointer (mouse, touch, or pen) moves.
 *
 * @example
 * ```typescript
 * const event: PointerMovedEvent = {
 *   event: 'on-move',
 *   data: {
 *     windowId: 1,
 *     pointerType: 'mouse',
 *     pointerId: 0,
 *     position: { x: 640, y: 360 }
 *   }
 * };
 * ```
 */
export interface PointerMovedEvent {
  event: 'on-move';
  data: {
    windowId: number;
    pointerType: PointerType;
    pointerId: number;
    position: Vector2;
  };
}

export interface PointerEnteredEvent {
  event: 'on-enter';
  data: {
    windowId: number;
    pointerType: PointerType;
    pointerId: number;
  };
}

export interface PointerLeftEvent {
  event: 'on-leave';
  data: {
    windowId: number;
    pointerType: PointerType;
    pointerId: number;
  };
}

/**
 * Event fired when a pointer button (mouse button) is pressed or released.
 *
 * @example
 * ```typescript
 * const event: PointerButtonEvent = {
 *   event: 'on-button',
 *   data: {
 *     windowId: 1,
 *     pointerType: 'mouse',
 *     pointerId: 0,
 *     button: 'left',
 *     state: 'pressed',
 *     position: { x: 100, y: 200 }
 *   }
 * };
 * ```
 */
export interface PointerButtonEvent {
  event: 'on-button';
  data: {
    windowId: number;
    pointerType: PointerType;
    pointerId: number;
    button: MouseButton;
    state: ElementState;
    position: Vector2;
  };
}

export interface PointerScrollEvent {
  event: 'on-scroll';
  data: {
    windowId: number;
    delta: ScrollDelta;
    phase: TouchPhase;
  };
}

export interface PointerTouchEvent {
  event: 'on-touch';
  data: {
    windowId: number;
    pointerId: number;
    phase: TouchPhase;
    position: Vector2;
    pressure: number | null;
  };
}

export interface PointerPinchGestureEvent {
  event: 'on-pinch-gesture';
  data: {
    windowId: number;
    delta: number;
    phase: TouchPhase;
  };
}

export interface PointerPanGestureEvent {
  event: 'on-pan-gesture';
  data: {
    windowId: number;
    delta: Vector2;
    phase: TouchPhase;
  };
}

export interface PointerRotationGestureEvent {
  event: 'on-rotation-gesture';
  data: {
    windowId: number;
    delta: number;
    phase: TouchPhase;
  };
}

export interface PointerDoubleTapGestureEvent {
  event: 'on-double-tap-gesture';
  data: { windowId: number };
}

export type PointerEvent =
  | PointerMovedEvent
  | PointerEnteredEvent
  | PointerLeftEvent
  | PointerButtonEvent
  | PointerScrollEvent
  | PointerTouchEvent
  | PointerPinchGestureEvent
  | PointerPanGestureEvent
  | PointerRotationGestureEvent
  | PointerDoubleTapGestureEvent;

File: src/core/input/events/pointer.rs
use glam::Vec2;
use serde::{Deserialize, Serialize};

use super::common::{ElementState, TouchPhase};

/// Mouse scroll delta type
#[derive(Debug, Clone, Copy, PartialEq, Deserialize, Serialize)]
#[serde(tag = "type", content = "value", rename_all = "kebab-case")]
pub enum ScrollDelta {
    /// Line-based scrolling (traditional mouse wheel)
    Line(Vec2),
    /// Pixel-based scrolling (touchpad)
    Pixel(Vec2),
}

/// Pointer (Mouse/Touch) events - unified for both input types
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "event", content = "data", rename_all = "kebab-case")]
pub enum PointerEvent {
    /// Pointer moved
    #[serde(rename_all = "camelCase")]
    OnMove {
        window_id: u32,
        pointer_type: u32,
        pointer_id: u64,
        position: Vec2,
    },

    /// Pointer entered window area
    #[serde(rename_all = "camelCase")]
    OnEnter {
        window_id: u32,
        pointer_type: u32,
        pointer_id: u64,
    },

    /// Pointer left window area
    #[serde(rename_all = "camelCase")]
    OnLeave {
        window_id: u32,
        pointer_type: u32,
        pointer_id: u64,
    },

    /// Pointer button pressed/released (mouse) or touch started/ended
    #[serde(rename_all = "camelCase")]
    OnButton {
        window_id: u32,
        pointer_type: u32,
        pointer_id: u64,
        button: u32,
        state: ElementState,
        position: Vec2,
    },

    /// Mouse wheel/touchpad scroll
    #[serde(rename_all = "camelCase")]
    OnScroll {
        window_id: u32,
        delta: ScrollDelta,
        phase: TouchPhase,
    },

    /// Touch event with pressure and additional info
    #[serde(rename_all = "camelCase")]
    OnTouch {
        window_id: u32,
        pointer_id: u64,
        phase: TouchPhase,
        position: Vec2,
        pressure: Option<f32>,
    },

    /// Pinch gesture (zoom)
    #[serde(rename_all = "camelCase")]
    OnPinchGesture {
        window_id: u32,
        delta: f64,
        phase: TouchPhase,
    },

    /// Pan gesture
    #[serde(rename_all = "camelCase")]
    OnPanGesture {
        window_id: u32,
        delta: Vec2,
        phase: TouchPhase,
    },

    /// Rotation gesture
    #[serde(rename_all = "camelCase")]
    OnRotationGesture {
        window_id: u32,
        delta: f32,
        phase: TouchPhase,
    },

    /// Double tap gesture
    #[serde(rename_all = "camelCase")]
    OnDoubleTapGesture { window_id: u32 },
}
